bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/commands/CommandImagePersistence.java

```diff
@@ -126,6 +131,58 @@ private static void readImagesFromRegistry(final IConfigurationElement[] configu
 				"Warnings while parsing the images from the 'org.eclipse.ui.commandImages' extension point."); //$NON-NLS-1$
 	}
 
+	private static void readImagesFromApplicationModel(final CommandImageManager commandImageManager) {
+
+		MApplication application = E4Workbench.getServiceContext().get(MApplication.class);
+
+		final List<IStatus> warningsToLog = new ArrayList<>();
+
+		for (MCommand command : application.getCommands()) {
+
+			if (command.getIconURI() != null) {
+				try {
+					registerImage(command, CommandImageManager.TYPE_DEFAULT, commandImageManager, command.getIconURI());
+				} catch (MalformedURLException e) {
+					addWarning(warningsToLog, "Malformed Icon Url: " + e.getMessage(), null, command.getElementId(), //$NON-NLS-1$
+							"iconUri", //$NON-NLS-1$
+							command.getIconURI());
+				}
+			}
+			if (command.getPersistedState().containsKey(E4_COMMAND_IMAGE_KEY_DISABLED)) {
```

Comment:
i wonder if that should also be part of the model then... I always missed specifying disabled/hover/pressed/... in other places as well!

Reply:
It's not directly part of the model, these are still handled over the persisted state, this is also possible for Tool Items and only necessary as a fallback when the automatic generation is not creating the desired result. 

If you want, I can remove the functionality, but i would like to keep it to make it feature complete to e3

Reply:
if it is in e3 we should also have it in e4, eg. `iconUriHover` and `iconUriDisabled`, maybe it would even be usefull to have these in a base model-class ... I think also menuitems and toolbar buttons can then get this in a more direct way.

Reply:
I think this should then not be part of this pr and discussed separately.
This should then also be handled differently in the E4 Model Editor to make it clear to the user that this is optional

Reply:
When exactly are `E4_COMMAND_IMAGE_KEY_DISABLED` / `E4_COMMAND_IMAGE_KEY_HOVER` set in the current implementation?

I agree with @laeubi : For symmetry with https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/extension-points/org_eclipse_ui_commandImages.html?cp=2_1_1_144

couldn't we instead simply also add the attributes `commandIconDisabledURI` and `commandIconHoverURI` directly to the model right now?

Reply:
@sratz
They are added manually in the persisted state.
Yes we could add them right now, but afaik there is also a E3 way to add disabled images to toolbar/menu items https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fextension-points%2Forg_eclipse_ui_menus.html&cp%3D2_1_1_144
which are also not directly supported in the E4 Model, only using the transient data.

I would like to separate those topics, either remove the disabled icon for now from this PR or leave them as is. And have a separate discussion on the disabled icons, if we want to have them in the E4 Model, how this should be handled in the editor and in the implementations.
For most cases the disabled icons are not relevant as they are generated at runtime when needed.

Reply:
> They are added manually in the persisted state.

I'm not sure I fully understand then (but I am not that familiar with E4 so bear with me): Who adds them manually? I see no code that _writes_ those attributes, only _reads_. Would that mean those constants are effectively API for consumer code?

> I think the best is to split this up and start with the bare minimum (a single icon) and then go for disabled / hover in a separate PR (that probably will fix this for Toolbars as well).

Sounds reasonable.

Reply:
@sratz basically yes, the E4 Model contains a Key Value Map for each element, and those constants would then be a key for that map. The consumer could add them in code or in their fragment.xmi file.

@laeubi @sratz I removed the disabled/hover image for now.